### PR TITLE
Extract IMA hashes from local and remote repositories

### PR DESCRIPTION
**Pull request by aplanas**

### CODE REVIEW

**Comment by aplanas** [comment]
Some improvements:

- Remove the Go idioms
- Promote the script as another Keylime cmdline tools

File: scripts/create_policy [hunk 1]
```diff
@@ -160,11 +189,183 @@ def process_signature_verification_keys(verification_keys, policy):
...
 
 
+def analyze_rpm_pkg(pkg):
+    """Analyze a single RPM package."""
+    ts = rpm.TransactionSet()
+    ts.setVSFlags(rpm.RPMVSF_MASK_NOSIGNATURES | rpm.RPMVSF_MASK_NODIGESTS)
+
+    with open(pkg, "rb") as f:
+        hdr = ts.hdrFromFdno(f)
+    # TODO - How are symbolic links managed with IMA
```
Comment (stefanberger) [inline scripts/create_policy:199]: if `ls` is a symbolic link to `busybox` then busybox will be logged an measured

File: scripts/create_policy [hunk 2]
```diff
@@ -160,11 +189,183 @@ def process_signature_verification_keys(verification_keys, policy):
...
+    while True:
+        with tempfile.TemporaryFile() as f:
+            range_ = f"{len(blob)}-{len(blob) + chunk_size - 1}"
+            req = urllib.request.Request(url, headers={"Range": f"bytes={range_}"})
+            with urllib.request.urlopen(req) as resp:
+                blob += resp.read()
+
+            f.write(blob)
+            f.seek(0)
+
```
Comment (stefanberger) [inline scripts/create_policy:231]: you should be able to call `analyze_rpm_pkg` on this package now ?

File: scripts/create_policy [hunk 3]
```diff
@@ -242,19 +446,29 @@ def main():
...
-    # Add the digests map either from the allowlist, if given, or the IMA measurement list.
+    if (args.local_repo or args.remote_repo) and not HAS_RPM:
+        print('To analyze RPM repositories the "rpm" Python module is required', file=sys.stderr)
+        sys.exit(1)
+
+    # Add the digests map either from the allowlist, a repo, or the
+    # IMA measurement list.
     if args.allowlist:
-        policy["digests"], ret = process_flat_allowlist(args.allowlist, policy["digests"])
+        policy["digests"], ret = process_flat_allowlist(args.allowlist)
```
Comment (stefanberger) [inline scripts/create_policy:464]: This is not the same as before... If the tool read some sort of base policy then this here should merge new values into it - at least that's what was the original intention. Now it's a replacement of the contents of the base policy. Both operations are probably valid, so this is a bit  'tricky'  on how to give users this type of control over it without the tool becoming confusing to use.

File: scripts/create_policy [hunk 4]
```diff
@@ -242,19 +446,29 @@ def main():
...
+    elif args.local_repo:
+        policy["digests"], ret = analize_local_repo(args.local_repo)
+    elif args.remote_repo:
+        policy["digests"], ret = analize_remote_repo(args.remote_repo)
     elif not args.no_hashes:
-        policy["digests"], ret = get_hashes_from_measurement_list(args.ima_measurement_list, policy["digests"])
+        policy["digests"], ret = get_hashes_from_measurement_list(args.ima_measurement_list)
```
Comment (stefanberger) [inline scripts/create_policy:470]: same comment here in regards to replacement versus merging ...

File: scripts/create_policy [hunk 5]
```diff
@@ -40,14 +40,16 @@ def process_flat_allowlist(allowlist_file):
                     continue
                 pieces = line.split(None, 1)
                 if not len(pieces) == 2:
-                    print(f"Skipping line that was split into {len(pieces)} parts, " "expected 2: {line}")
```
Comment (stefanberger) [inline scripts/create_policy:42]: commit text typo: emission & break

Comment (stefanberger) [reply scripts/create_policy:42]: Otherwise this patch LGTM

File: scripts/create_policy (hunk 1 above)
Comment (aplanas) [reply scripts/create_policy:199]: Thanks, that simplify the problem. I will update the doc.

File: scripts/create_policy (hunk 3 above)
Comment (aplanas) [reply scripts/create_policy:464]: Reverting this then. Thanks!

File: scripts/create_policy (hunk 5 above)
Comment (aplanas) [reply scripts/create_policy:42]: Fixed

**Review by maugustosilva — APPROVED** [review]

**Comment by maugustosilva** [comment]
Please squash the commits and LGTM

**Comment by aplanas** [comment]
> Please squash the commits

I would prefer to not squash them. They are affecting different files, and represent different stages of the feature. But if there is a consensus then they can be squashed.

**Comment by maugustosilva** [comment]
Well, the "standard operating procedure" for all PRs here is to have all commits squashed. I'd rather let you decide how to squash and how to work the unified commit message than do it via the "Squash and merge" button on the Web UI :-)

**Comment by ansasaki** [comment]
I didn't know about this "always squash" policy. I personally prefer to have separate commits reasonably self contained that make specific and easy-to-review changes. It is a nightmare later to search through the commit history to find where a specific change was made if all commits are squashed.

**Comment by maugustosilva** [comment]
Will have a short meeting with @mpeters and others maintainers soon, and will clarify this point. I have a preference for squashing (a self-contained change), but don't see it as crucial.

**Comment by aplanas** [comment]
I was not aware of the policy neither. In fact I found it a mistake, and indeed there are recent PR that have multiple commits:

(6 commits, 2 weeks ago) https://github.com/keylime/keylime/pull/1291
(7 commits, 2 weeks ago) https://github.com/keylime/keylime/pull/1287

IMHO we should drop the policy, also because the mentioned reason exposed by @ansasaki, each commit can explain the rationale of the change, and conflating 8 commit are as not deeply related can be a pain when looking for an explanation.

Since long ago github also relate the commit into the pull request, so it is very easy when doing "blame" to get the single commit and from there the pull request that contains the extended discussion.

**Comment by maugustosilva** [comment]
@aplanas aha, so it was not being enforced uniformly already! Ok, will have a quick chat (with @mpeters and others) and make sure we have a unified policy on it.

**Comment by maugustosilva** [comment]
@aplanas @THS-on Ok, question settled. We might ask for it (squashing) but this will not be mandatory any longer. There are scenarios where having a PR with multiple commits indeed makes sense... if the reviewer decides that there are too many "fragmented" commits "polluting" the PR he can ask for a squashing.

**Comment by aplanas** [comment]
Good news: this has been merged https://github.com/rpm-software-management/createrepo_c/pull/316 (later the metadata has been renamed to filelist-ext.xml)

This imply that the second algorithm for remote repositories can be adapted, and if this file is present the performance will be massively improved.

**Pull request merged**